src/Arcus.Messaging.Pumps.Abstractions/MessagePump.cs

```diff
@@ -0,0 +1,124 @@
+        protected virtual Task HandleReceiveExceptionAsync(Exception receiveException)
+        {
+            Logger.LogError(receiveException, "Unable to process message");
+            return Task.CompletedTask;
+        }
+
+        /// <summary>
+        ///     Deserializes a raw JSON message body
+        /// </summary>
+        /// <param name="rawMessageBody">Raw message body to deserialize</param>
+        /// <param name="messageContext">Context concerning the message</param>
+        /// <returns>Deserialized message</returns>
+        protected virtual TMessage DeserializeJsonMessageBody(byte[] rawMessageBody, MessageContext messageContext)
+        {
+            Encoding encoding = DetermineMessageEncoding(messageContext);
+            string serializedMessageBody = encoding.GetString(rawMessageBody);
+
+            TMessage messageBody = JsonConvert.DeserializeObject<TMessage>(serializedMessageBody);
+            if (messageBody == null)
+            {
+                Logger.LogError("Unable to deserialize to message contract {ContractName} for message {MessageBody}",
+                    typeof(TMessage), rawMessageBody);
+            }
+
+            return messageBody;
+        }
+
+        /// <summary>
+        ///     Process a new message that was received
+        /// </summary>
+        /// <param name="message">Message that was received</param>
+        /// <param name="messageContext">Context providing more information concerning the processing</param>
+        /// <param name="correlationInfo">
+        ///     Information concerning correlation of telemetry & processes by using a variety of unique
+        ///     identifiers
+        /// </param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        protected abstract Task ProcessMessageAsync(TMessage message, TMessageContext messageContext,
+            MessageCorrelationInfo correlationInfo, CancellationToken cancellationToken);
+
+        /// <summary>
+        ///     Determines the encoding used for a given message
+        /// </summary>
+        /// <remarks>If no encoding was specified, UTF-8 will be used by default</remarks>
+        /// <param name="messageContext">Context concerning the message</param>
+        /// <returns>Encoding that was used for the message body</returns>
+        protected Encoding DetermineMessageEncoding(MessageContext messageContext)
+        {
+            var encoding = Encoding.UTF8;
+            if (messageContext.Properties.TryGetValue(PropertyNames.Encoding, out object annotatedEncoding))
+                try
+                {
+                    encoding = Encoding.GetEncoding(annotatedEncoding.ToString());
+                }
+                catch (Exception ex)
+                {
+                    Logger.LogCritical(ex,
+                        $"Unable to determine encoding with name '{{Encoding}}'. Falling back to {encoding.WebName}.",
```

Comment:
Both string concatenation and formatting is used here. May want to make the `encoding.WebName` also a formatting input.
And the remove the unnecessary extra `{`` in the `{{Encoding}}` formatting input?

Reply:
Good catch! I've made them both use the formatting approach because otherwise we don't have structured logs anymore